pkg/storegateway/bucket.go

```diff
@@ -1903,14 +1899,38 @@ func (s *bucketBlockSet) forEach(fn func(b *bucketBlock)) {
 	})
 }
 
-func (s *bucketBlockSet) blockULIDs() []ulid.ULID {
+// closeAll closes all blocks in the set and returns all encountered errors after trying all blocks.
+func (s *bucketBlockSet) closeAll() error {
+	errs := multierror.New()
+	s.blockSet.Range(func(_, val any) bool {
+		errs.Add(val.(*bucketBlock).Close())
```

Comment:
To double-check, I got the point in the description right: `bucketBlock.Close()` is expected to be idempotent, right? 

It seems that we can (potentially) call it from several places. Also, looks like the change in `*LazyBinaryReader.Close()` below (about `close(done)` channel) introduces a potential panic, if multiple goroutines tries to close the `indexHeaderReader`. Is this not a problem here?

Reply:
sorry for the delay @narqo 

> To double-check that I got the points from the description right: `bucketBlock.Close()` is expected to be idempotent, right?

I didn't check about this. So I had to review it again. I don't think `bucketBlock.Close()` is expected to be idempotent. There are multiple code paths that can close the block, but I believe they are mutually exclusive:

1. [when the block is being removed from disk via `BucketStore.removeBlock()`](https://github.com/grafana/mimir/blob/a54091cedb143c0f6758c2df087ad2b564699d7f/pkg/storegateway/bucket.go#L510-L522) (scenario 2. and 4. from the description). In this case the block is also removed from the block set - `s.blockSet.remove(id)`, so we can't call `bucketBlock.Close()` multiple times even if  `BucketStore.removeBlock()` is called multiple times
2. [when the BucketStore is being shut down, so all blocks need to be closed via `bucketBlockSet.closeAll()`](https://github.com/grafana/mimir/blob/a54091cedb143c0f6758c2df087ad2b564699d7f/pkg/storegateway/bucket.go#L1906) (scenario 3. in the description). This runs once in the process' lifetime.

1\. can never run after 2. and 2. cannot discover the block if 1. has previously ran. 1. and 2. can also never run concurrently. 1. can only happen as a result of the periodic `StoreGateway.running > BucketStores.SyncBlocks > BucketStore.SyncBlocks`. 2. can only happen as a result of `StoreGateway.stopping > BucketStores.stopBucketStores > BucketStore.stop`. `StoreGateway.running` and `StoreGateway.stopping` can't be running concurrently.


It's a long-winded answer, but I hope it was clear. Happy to jump on a call or chat on slack

Reply:
> 1. ...
> 2. ...

the way I discovered this is by renaming `bucketBlock.Close()` to `bucketBlock.CloseBlock()` in my IDE and finding all usages. This prevents the name conflict with `io.Closer` and shows these two usages only.